### PR TITLE
[16.0][MIG] stock_cycle_count

### PR DESCRIPTION
**Pull request by ArnauCForgeFlow**
Migration to version 16.0

### CODE REVIEW

**Comment by LoisRForgeFlow** [comment]
@ArnauCForgeFlow Could you rebase this PR?

while testing I found an issue that was solved in another PR (https://github.com/OCA/stock-logistics-warehouse/pull/2048)

```
    res = super()._apply_inventory()
  File "/mnt/data/odoo-addons-dir/stock_inventory/models/stock_quant.py", line 54, in _apply_inventory
    if adjustment and self.env.company.stock_inventory_auto_complete:
UnboundLocalError: local variable 'adjustment' referenced before assignment
```

**Review by LoisRForgeFlow — COMMENTED** [review]
Tested in runboat, an error raises when validating a discrepancy:

image

image

File: stock_cycle_count/models/stock_cycle_count.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+            "cycle_count_id": self.id,
+            "location_ids": [(4, self.location_id.id)],
+            "exclude_sublocation": True,
+            "responsible_id": self.responsible_id.id,
+        }
+
+    @api.model
+    def create(self, vals):
+        vals["name"] = self.env["ir.sequence"].next_by_code("stock.cycle.count") or ""
+        return super(StockCycleCount, self).create(vals)
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count.py]: You can adopt short notation
```suggestion
        return super().create(vals)
```

File: stock_cycle_count/models/stock_cycle_count.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        return {
+            "name": "INV/{}".format(self.name),
+            "cycle_count_id": self.id,
+            "location_ids": [(4, self.location_id.id)],
+            "exclude_sublocation": True,
+            "responsible_id": self.responsible_id.id,
+        }
+
+    @api.model
+    def create(self, vals):
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count.py]: You should use multi create.

File: stock_cycle_count/models/stock_cycle_count.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        vals["name"] = self.env["ir.sequence"].next_by_code("stock.cycle.count") or ""
+        return super(StockCycleCount, self).create(vals)
+
+    def action_create_inventory_adjustment(self):
+        if any([s != "draft" for s in self.mapped("state")]):
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count.py]:
```suggestion
        if any([state != "draft" for state in self.mapped("state")]):
```

Explicit is better

File: stock_cycle_count/models/stock_cycle_count.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            raise UserError(_("You can only confirm cycle counts in state 'Planned'."))
+        for rec in self:
+            data = rec._prepare_inventory_adjustment()
+            inv = self.env["stock.inventory"].create(data)
+            if self.company_id.auto_start_inventory_from_cycle_count:
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count.py]: This is not correct

```suggestion
            if rec.company_id.auto_start_inventory_from_cycle_count:
```

File: stock_cycle_count/models/stock_cycle_count.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+                inv.prefill_counted_quantity = (
+                    self.company_id.inventory_adjustment_counted_quantities
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count.py]: Same as above
```suggestion
                    rec.company_id.inventory_adjustment_counted_quantities
```

File: stock_cycle_count/models/stock_cycle_count.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                if inv.prefill_counted_quantity == "zero":
+                    inv.stock_quant_ids.write({"inventory_quantity": 0})
+                else:
+                    for quant in inv.stock_quant_ids:
+                        quant.write({"inventory_quantity": quant.quantity})
+        self.write({"state": "open"})
+        return True
+
+    def action_view_inventory(self):
+        action = self.env.ref(
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count.py]: Use _for_xml_id()

File: stock_cycle_count/models/stock_cycle_count_rule.py [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+    def _onchange_locaton_ids(self):
+        """Get the warehouses for the selected locations."""
+        wh_ids = []
+        for loc in self.location_ids:
+            wh_ids.append(loc.warehouse_id.id)
+        wh_ids = list(set(wh_ids))
+        wh_ids = [wh for wh in wh_ids if wh]
+        self.warehouse_ids = self.env["stock.warehouse"].browse(wh_ids)
+
+    name = fields.Char(required=True)
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count_rule.py]: Put all fields definitions on top

File: stock_cycle_count/models/stock_cycle_count_rule.py [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+                    _(
+                        "You cannot define a negative or null number of counts "
+                        "per period."
+                    )
+                )
+            if rec.periodic_count_period < 0:
+                raise ValidationError(_("You cannot define a negative period."))
+
+    @api.onchange("location_ids")
+    def _onchange_locaton_ids(self):
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count_rule.py]: Change warehouse_ids field to compute/store/readonly=False style and change this method to compute

File: stock_cycle_count/models/stock_cycle_count_rule.py [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+            )
+            if latest_inventory_date:
+                try:
+                    period = self.periodic_count_period / self.periodic_qty_per_period
+                    next_date = fields.Datetime.from_string(
+                        latest_inventory_date
+                    ) + timedelta(days=period)
+                    if next_date < datetime.today():
+                        next_date = datetime.today()
+                except Exception as e:
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count_rule.py]: What exception do you try to catch ? You should be more explicit.

File: stock_cycle_count/models/stock_cycle_count_rule.py [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+        price = move._get_price_unit()
+        turnover = move.product_uom_qty * price
+        return turnover
+
+    @api.model
+    def _compute_rule_turnover(self, locs):
+        cycle_counts = []
+        for loc in locs:
+            last_inventories = (
+                self.env["stock.inventory"]
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count_rule.py]: Put this search outside the loop.

File: stock_cycle_count/models/stock_cycle_count_rule.py [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+                .search(
+                    [
+                        ("location_ids", "in", [loc.id]),
+                        ("state", "in", ["confirm", "done", "draft"]),
+                    ]
+                )
+                .mapped("date")
+            )
+            if last_inventories:
+                latest_inventory = sorted(last_inventories, reverse=True)[0]
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count_rule.py:242]:
```suggestion
                from odoo.fields import first
                latest_inventory = first(sorted(last_inventories, reverse=True))
```

File: stock_cycle_count/models/stock_cycle_count_rule.py [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+                    total_turnover = 0.0
+                    for m in moves:
+                        turnover = self._compute_turnover(m)
+                        total_turnover += turnover
+                    try:
+                        if total_turnover > self.turnover_inventory_value_threshold:
+                            next_date = datetime.today()
+                            cycle_count = self._propose_cycle_count(next_date, loc)
+                            cycle_counts.append(cycle_count)
+                    except Exception as e:
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_cycle_count_rule.py]: Same as above

File: stock_cycle_count/models/stock_inventory.py [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+                inv.stock_quant_ids.mapped(lambda x: abs(x.inventory_diff_quantity))
+            )
+            if theoretical:
+                inv.inventory_accuracy = max(
+                    PERCENT * (theoretical - abs_discrepancy) / theoretical, 0.0
+                )
+            if not inv.stock_quant_ids and inv.state == "done":
+                inv.inventory_accuracy = PERCENT
+
+    prefill_counted_quantity = fields.Selection(
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_inventory.py:14]: Please put fields definitions on top

File: stock_cycle_count/models/stock_inventory.py [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+                self.exclude_sublocation
+                or (len(self.location_ids) == 1 and not self.location_ids[0].child_ids)
+            )
+        ):
+            candidate.state = "open"
+            self.write({"cycle_count_id": candidate.id, "exclude_sublocation": True})
+        return True
+
+    def action_state_to_done(self):
+        res = super(StockInventory, self).action_state_to_done()
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_inventory.py]:
```suggestion
        res = super().action_state_to_done()
```

File: stock_cycle_count/models/stock_inventory.py [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+        self._update_cycle_state()
+        return res
+
+    def action_force_done(self):
+        res = super(StockInventory, self).action_force_done()
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_inventory.py]:
```suggestion
        res = super().action_force_done()
```

File: stock_cycle_count/models/stock_inventory.py [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+        self._update_cycle_state()
+        return res
+
+    @api.model
+    def create(self, vals):
+        res = super().create(vals)
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_inventory.py]: Change this to multi style

File: stock_cycle_count/models/stock_inventory.py [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+        "company_id",
+        "exclude_sublocation",
+    )
+    def _check_cycle_count_consistency(self):
+        for rec in self.filtered(lambda r: r.cycle_count_id):
+            is_consistent, msg = rec._is_consistent_with_cycle_count()
+            if not is_consistent:
+                raise ValidationError(
+                    _(
+                        "The Inventory Adjustment is inconsistent with the Cycle Count:\n%s"
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_inventory.py]:
```suggestion
                        "The Inventory Adjustment is inconsistent with the Cycle Count:\n%(message)s", message=msg
```

File: stock_cycle_count/models/stock_location.py [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+except (ImportError, IOError) as err:
+    _logger.debug(err)
+
+
+class StockLocation(models.Model):
+    _inherit = "stock.location"
+
+    def _compute_loc_accuracy(self):
+        for rec in self:
+            history = self.env["stock.inventory"].search(
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_location.py]: IMHO this should be refined to improve performances.

File: stock_cycle_count/models/stock_location.py [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+                        history[: wh.counts_for_accuracy_qty].mapped(
+                            "inventory_accuracy"
+                        )
+                    )
+                else:
+                    rec.loc_accuracy = mean(history.mapped("inventory_accuracy"))
+            else:
+                rec.loc_accuracy = 0
+
+    zero_confirmation_disabled = fields.Boolean(
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_location.py:24]: Put fields defintions on top

File: stock_cycle_count/models/stock_location.py [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+        self.ensure_one()
+        domain = [("location_id", "=", self.id), ("quantity", ">", 0.0)]
+        return domain
+
+    def check_zero_confirmation(self):
+        for rec in self:
+            if not rec.zero_confirmation_disabled:
+                wh = rec.warehouse_id
+                rule_model = self.env["stock.cycle.count.rule"]
+                zero_rule = rule_model.search(
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_location.py]: Retrieve all zero rules per warehouse before the loop

File: stock_cycle_count/models/stock_location.py [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+                        rec.create_zero_confirmation_cycle_count()
+
+    def create_zero_confirmation_cycle_count(self):
+        self.ensure_one()
+        date = datetime.today().strftime(DEFAULT_SERVER_DATETIME_FORMAT)
+        wh_id = self.warehouse_id.id
+        date_horizon = self.warehouse_id.get_horizon_date().strftime(
+            DEFAULT_SERVER_DATETIME_FORMAT
+        )
+        counts_planned = self.env["stock.cycle.count"].search(
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_location.py:95]: Should maybe become a field of the location (e.g. 'planned_cycle_count_ids') ?

File: stock_cycle_count/models/stock_warehouse.py [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+        self.ensure_one()
+        date = datetime.today()
+        delta = timedelta(self.cycle_count_planning_horizon)
+        date_horizon = date + delta
+        return date_horizon
+
+    @api.model
+    def _get_cycle_count_locations_search_domain(self, parent):
+        domain = [
+            ("parent_path", "=like", parent.parent_path + "%"),
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_warehouse.py]: Why not using 'child of' operator ?

File: stock_cycle_count/models/stock_warehouse.py [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+            "location_id": cycle_count_proposed["location"].id,
+            "cycle_count_rule_id": cycle_count_proposed["rule_type"].id,
+            "state": "draft",
+        }
+
+    def action_compute_cycle_count_rules(self):
+        """Apply the rule in all the sublocations of a given warehouse(s) and
+        returns a list with required dates for the cycle count of each
+        location"""
+        for rec in self:
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_warehouse.py:83]: Maybe this compute function is too long and can be decoupled in some other.

File: stock_cycle_count/models/stock_warehouse.py [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+                    delta = (
+                        fields.Datetime.from_string(cycle_count_proposed["date"])
+                        - datetime.today()
+                    )
+                    if (
+                        not existing_cycle_counts
+                        and delta.days < rec.cycle_count_planning_horizon
+                    ):
+                        cc_vals = self._prepare_cycle_count(cycle_count_proposed)
+                        self.env["stock.cycle.count"].create(cc_vals)
```
Comment (rousseldenis) [inline stock_cycle_count/models/stock_warehouse.py]: You should prepare a list of vals and use multi create

File: stock_cycle_count/reports/report_stock_location_accuracy.py [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+            ("filter", "=", "none"),
+            ("state", "=", "done"),
+        ]
+
+    @api.model
+    def _get_location_data(self, locations):
+        data = dict()
+        inventory_obj = self.env["stock.inventory"]
+        for loc in locations:
+            counts = inventory_obj.search(self._get_inventory_domain(loc.id))
```
Comment (rousseldenis) [inline stock_cycle_count/reports/report_stock_location_accuracy.py]: You should change the search to use multi recordset, call search() once, then split per location.

File: stock_cycle_count/tests/test_stock_cycle_count.py [hunk 26]
```diff
@@ -0,0 +1,340 @@
...
+#   (http://www.forgeflow.com)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+from datetime import datetime, timedelta
+
+from odoo.exceptions import AccessError, ValidationError
+from odoo.tests import common
+
+
+class TestStockCycleCount(common.TransactionCase):
+    def setUp(self):
```
Comment (rousseldenis) [inline stock_cycle_count/tests/test_stock_cycle_count.py]: Change this to classmethod

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review.

At least some features should be refined for performance reasons

**Comment by LoisRForgeFlow** [comment]
@rousseldenis Thanks for your review, indeed they are good remarks. @ArnauCForgeFlow could you apply those fixes in a separated commit? This way it would be easier to cherry pick and back port it to v15.

**Comment by ArnauCForgeFlow** [comment]
@LoisRForgeFlow Sure!

**Comment by ArnauCForgeFlow** [comment]
Hi @rousseldenis I've already made the changes you asked for in the review. Could you please take a look and let me know if everything's okay?

File: stock_cycle_count/models/stock_location.py (hunk 18 above)
Comment (rousseldenis) [reply stock_cycle_count/models/stock_location.py]: You didn't solve this. You should put the search before the loop with 'location_ids', 'in', self.ids.

You can then add here history(_ids).filtered_domain([("location_ids", "in", rec.id)])

File: stock_cycle_count/models/stock_location.py (hunk 19 above)
Comment (rousseldenis) [reply stock_cycle_count/models/stock_location.py:24]: Haven't solved this

File: stock_cycle_count/models/stock_warehouse.py (hunk 22 above)
Comment (rousseldenis) [reply stock_cycle_count/models/stock_warehouse.py]: Question not answered :smiley:

File: stock_cycle_count/models/stock_location.py (hunk 18 above)
Comment (ArnauCForgeFlow) [reply stock_cycle_count/models/stock_location.py]: Done

File: stock_cycle_count/models/stock_location.py (hunk 19 above)
Comment (ArnauCForgeFlow) [reply stock_cycle_count/models/stock_location.py:24]: Done

File: stock_cycle_count/models/stock_warehouse.py (hunk 22 above)
Comment (ArnauCForgeFlow) [reply stock_cycle_count/models/stock_warehouse.py]: I hadn't considered that, but it seems that using 'child of' is cleaner, changed.

**Review by rousseldenis — APPROVED** [review]

**Review by LoisRForgeFlow — APPROVED** [review]
/ocabot merge nobump

**Comment by xaviedoanhduy** [comment]
Hello, thanks a lot for your work on migrating this module! 

While reviewing the code, I noticed that the module still inherits and relies heavily on `stock.inventory`, despite the fact that in `15.0` Odoo has merged much of its logic into `stock.quant`, as per this upstream change: https://github.com/odoo/odoo/pull/68409

I'm wondering if this might lead to inconsistencies or a loss of features, as the module doesn’t seem to take that upstream refactor into account. Would you recommend keeping the old `stock.inventory`-based logic, or would it make more sense to adapt it to the new `stock.quant` flow?

Thanks again! 

cc @LoisRForgeFlow, @ArnauCForgeFlow

**Pull request merged**